sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/MultimapUserState.java

```diff
@@ -0,0 +1,326 @@
+                .setClear(StateClearRequest.getDefaultInstance()),
+            new CompletableFuture<>());
+      }
+    } else if (!pendingRemoves.isEmpty()) {
+      Iterable<K> removeKeys = Iterables.filter(getPersistedKeys(), pendingRemoves::contains);
+      for (K key : removeKeys) {
+        beamFnStateClient.handle(
+            createUserStateRequest(key)
+                .toBuilder()
+                .setClear(StateClearRequest.getDefaultInstance()),
+            new CompletableFuture<>());
+      }
+    }
+
+    // Persist pending key-values
+    if (!pendingAdds.isEmpty()) {
+      for (K key : pendingAdds.keySet()) {
+        beamFnStateClient.handle(
+            createUserStateRequest(key)
+                .toBuilder()
+                .setAppend(
+                    StateAppendRequest.newBuilder().setData(encodeValues(pendingAdds.get(key)))),
+            new CompletableFuture<>());
+      }
+    }
+
+    // Update keys
+    if (isCleared || !pendingRemoves.isEmpty()) {
+      beamFnStateClient.handle(
+          keysStateRequest.toBuilder().setClear(StateClearRequest.getDefaultInstance()),
+          new CompletableFuture<>());
+      beamFnStateClient.handle(
+          keysStateRequest
+              .toBuilder()
+              .setAppend(StateAppendRequest.newBuilder().setData(encodeKeys(keys())).build()),
+          new CompletableFuture<>());
+    } else {
+      beamFnStateClient.handle(
+          keysStateRequest
+              .toBuilder()
+              .setAppend(StateAppendRequest.newBuilder().setData(encodeKeys(pendingKeys)).build()),
+          new CompletableFuture<>());
+    }
+    isClosed = true;
+  }
+
+  private ByteString encodeKeys(Iterable<K> keys) {
+    try {
+      ByteString.Output output = ByteString.newOutput();
+      for (K key : keys) {
+        keyCoder.encode(key, output);
+      }
+      return output.toByteString();
+    } catch (IOException e) {
+      throw new IllegalStateException(
+          String.format("Failed to encode keys for multimap user state id %s.", stateId), e);
+    }
+  }
+
+  private ByteString encodeValues(Iterable<V> values) {
+    try {
+      ByteString.Output output = ByteString.newOutput();
+      for (V value : values) {
+        valueCoder.encode(value, output);
+      }
+      return output.toByteString();
+    } catch (IOException e) {
+      throw new IllegalStateException(
+          String.format("Failed to encode values for multimap user state id %s.", stateId), e);
+    }
+  }
+
+  private StateRequest createUserStateRequest(K key) {
+    ByteString.Output keyStream = ByteString.newOutput();
+    try {
+      encodedKey.writeTo(keyStream);
+      keyCoder.encode(key, keyStream);
+    } catch (IOException e) {
+      throw new IllegalStateException(
+          String.format("Failed to encode key %s for multimap user state id %s.", key, stateId), e);
+    }
+
+    StateRequest.Builder request = StateRequest.newBuilder();
+    request
```

Comment:
Re-use `keysStateRequest` instead of setting this all again via `keysStateRequest.toBuilder()`